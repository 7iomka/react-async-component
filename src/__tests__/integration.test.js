import React from 'react'
import { renderToStaticMarkup } from 'react-dom/server'
import { mount } from 'enzyme'

import {
  AsyncComponentProvider,
  createContext,
  createAsyncComponent,
  withAsyncComponents,
  STATE_IDENTIFIER,
} from '../'

function Bob({ children }) {
  return <div>{children}</div>
}
Bob.propTypes = { children: React.PropTypes.node }
Bob.defaultProps = { children: null }

const AsyncBob = createAsyncComponent({
  resolve: () => new Promise(resolve => setTimeout(() => resolve(Bob), 10)),
  name: 'AsyncBob',
})

const AsyncBobTwo = createAsyncComponent({
  resolve: () => new Promise(resolve => setTimeout(() => resolve(Bob), 10)),
  name: 'AsyncBobTwo',
})

const AsyncBobThree = createAsyncComponent({
  resolve: () => new Promise(resolve => setTimeout(() => resolve(Bob), 10)),
  name: 'AsyncBobThree',
})

const DeferredAsyncBob = createAsyncComponent({
  resolve: () => new Promise(resolve => setTimeout(() => resolve(Bob), 10)),
  ssrMode: 'defer',
  name: 'DeferredAsyncBob',
})

const BoundaryAsyncBob = createAsyncComponent({
  resolve: () => new Promise(resolve => setTimeout(() => resolve(Bob), 10)),
  ssrMode: 'boundary',
  name: 'BoundaryAsyncBob',
})

const app = execContext => (
  <AsyncComponentProvider execContext={execContext}>
    <AsyncBob>
      <div>
        <AsyncBobTwo>
          <span>In Render.</span>
        </AsyncBobTwo>
        <DeferredAsyncBob>
          <span>In Defer.</span>
        </DeferredAsyncBob>
        <BoundaryAsyncBob>
          <span>
            In Boundary but outside an AsyncComponent, server render me!
          </span>
          <AsyncBobThree>
            <span>In Boundary - Do not server render me!</span>
          </AsyncBobThree>
        </BoundaryAsyncBob>
      </div>
    </AsyncBob>
  </AsyncComponentProvider>
)

describe('integration tests', () => {
  afterEach(() => {
    delete global.window[STATE_IDENTIFIER]
  })

  it('render server and client', () => {
    const windowTemp = global.window
    // we have to delete the window to emulate a server only environment
    delete global.window

    // "Server" side render...
    const serverContext = createContext()
    const serverApp = app(serverContext)
    return withAsyncComponents(serverApp)
      .then(() => {
        const serverString = renderToStaticMarkup(serverApp)
        expect(serverString).toMatchSnapshot()
        expect(serverContext.getState()).toMatchSnapshot()
        // Restore the window and attach the state to the "window" for the client
        global.window = windowTemp
        global.window[STATE_IDENTIFIER] = serverContext.getState()
        return serverString
      })
      .then((serverHTML) => {
        // "Client" side render...
        const clientContext = createContext()
        const clientApp = app(clientContext)
        return (
          withAsyncComponents(clientApp)
            .then(() => {
              const clientRenderWrapper = mount(clientApp)
              expect(clientRenderWrapper).toMatchSnapshot()
              expect(renderToStaticMarkup(clientApp)).toEqual(serverHTML)
              return clientRenderWrapper
            })
            // Now give the client side components time to resolve
            .then(
              clientRenderWrapper =>
                new Promise(resolve => setTimeout(() => resolve(clientRenderWrapper), 100)),
            )
            // Now a full render should have occured on client
            .then(clientRenderWrapper => expect(clientRenderWrapper).toMatchSnapshot())
        )
      })
  })

  it('a component only gets resolved once', () => {
    let resolveCount = 0

    const Foo = createAsyncComponent({
      resolve: () => {
        resolveCount += 1
        return () => <div>foo</div>
      },
    })

    withAsyncComponents(<Foo />).then(({ appWithAsyncComponents }) => {
      mount(appWithAsyncComponents)
      expect(resolveCount).toEqual(1)
      mount(appWithAsyncComponents)
      expect(resolveCount).toEqual(1)
    })
  })
})
