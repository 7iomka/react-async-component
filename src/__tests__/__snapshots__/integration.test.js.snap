// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration tests render server and client 1`] = `"<div><div><div><span>In Render.</span></div><div>DeferredAsyncBob Loading</div><div><span>In Boundary but outside an AsyncComponent, server render me!</span></div><div>This always errors</div></div></div>"`;

exports[`integration tests render server and client 2`] = `
Object {
  "errors": Object {
    "6": "This always errors",
  },
  "resolved": Object {
    "1": true,
    "2": true,
    "4": true,
  },
}
`;

exports[`integration tests render server and client 3`] = `
<AsyncComponentProvider
  execContext={
    Object {
      "getComponent": [Function],
      "getError": [Function],
      "getNextId": [Function],
      "getState": [Function],
      "registerComponent": [Function],
      "registerError": [Function],
    }
  }
>
  <AsyncBob>
    <Bob>
      <div>
        <div>
          <AsyncBobTwo>
            <Bob>
              <div>
                <span>
                  In Render.
                </span>
              </div>
            </Bob>
          </AsyncBobTwo>
          <DeferredAsyncBob>
            <LoadingComponent>
              <div>
                DeferredAsyncBob Loading
              </div>
            </LoadingComponent>
          </DeferredAsyncBob>
          <BoundaryAsyncBob>
            <Bob>
              <div>
                <span>
                  In Boundary but outside an AsyncComponent, server render me!
                </span>
                <AsyncBobThree />
              </div>
            </Bob>
          </BoundaryAsyncBob>
          <ErrorAsyncComponent>
            <ErrorComponent
              message="This always errors"
            >
              <div>
                This always errors
              </div>
            </ErrorComponent>
          </ErrorAsyncComponent>
        </div>
      </div>
    </Bob>
  </AsyncBob>
</AsyncComponentProvider>
`;

exports[`integration tests render server and client 4`] = `
<AsyncComponentProvider
  execContext={
    Object {
      "getComponent": [Function],
      "getError": [Function],
      "getNextId": [Function],
      "getState": [Function],
      "registerComponent": [Function],
      "registerError": [Function],
    }
  }
>
  <AsyncBob>
    <Bob>
      <div>
        <div>
          <AsyncBobTwo>
            <Bob>
              <div>
                <span>
                  In Render.
                </span>
              </div>
            </Bob>
          </AsyncBobTwo>
          <DeferredAsyncBob>
            <Bob>
              <div>
                <span>
                  In Defer.
                </span>
              </div>
            </Bob>
          </DeferredAsyncBob>
          <BoundaryAsyncBob>
            <Bob>
              <div>
                <span>
                  In Boundary but outside an AsyncComponent, server render me!
                </span>
                <AsyncBobThree>
                  <Bob>
                    <div>
                      <span>
                        In Boundary - Do not server render me!
                      </span>
                    </div>
                  </Bob>
                </AsyncBobThree>
              </div>
            </Bob>
          </BoundaryAsyncBob>
          <ErrorAsyncComponent>
            <ErrorComponent
              message="This always errors"
            >
              <div>
                This always errors
              </div>
            </ErrorComponent>
          </ErrorAsyncComponent>
        </div>
      </div>
    </Bob>
  </AsyncBob>
</AsyncComponentProvider>
`;

exports[`integration tests renders the ErrorComponent 1`] = `
<AsyncComponentProvider
  execContext={
    Object {
      "getComponent": [Function],
      "getError": [Function],
      "getNextId": [Function],
      "getState": [Function],
      "registerComponent": [Function],
      "registerError": [Function],
    }
  }
>
  <AsyncComponent>
    <ErrorComponent
      message="An error occurred"
    >
      <div>
        An error occurred
      </div>
    </ErrorComponent>
  </AsyncComponent>
</AsyncComponentProvider>
`;

exports[`integration tests renders the LoadingComponent 1`] = `
<AsyncComponentProvider
  execContext={
    Object {
      "getComponent": [Function],
      "getError": [Function],
      "getNextId": [Function],
      "getState": [Function],
      "registerComponent": [Function],
      "registerError": [Function],
    }
  }
>
  <AsyncComponent>
    <LoadingComponent>
      <div>
        Loading...
      </div>
    </LoadingComponent>
  </AsyncComponent>
</AsyncComponentProvider>
`;
